File: 2024/01.py
from collections import Counter

input = []
with open('01.txt', 'r') as f:
    input = list(map(int, f.read().split()))
part1_sum = sum([abs(pair[0] - pair[1])
                 for pair in zip(sorted(input[::2]), sorted(input[1::2]))])
print('Part 1:', part1_sum)
right_dict = Counter(input[1::2])
similiarity = sum([left_num * right_dict[left_num]
                   for left_num in input[::2]])
print('Part 2:', similiarity)

File: 2024/01.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func absDiffInt(x, y int) int {
	if x < y {
		return y - x
	}
	return x - y
}

func main() {
	in := `3   4
	4   3
	2   5
	1   3
	3   9
	3   3`
	rawIn, _ := os.ReadFile("01.txt")
	in = string(rawIn)
	lines := strings.Split(in, "\n")
	var leftNums []int
	var rightNums []int
	for _, line := range lines {
		nums := strings.Fields(line)
		if len(nums) == 0 {
			break
		}
		leftNum, _ := strconv.Atoi(nums[0])
		rightNum, _ := strconv.Atoi(nums[1])
		leftNums = append(leftNums, leftNum)
		rightNums = append(rightNums, rightNum)
	}
	sort.Ints(leftNums)
	sort.Ints(rightNums)
	sum := 0
	for i := 0; i < len(leftNums); i++ {
		sum += absDiffInt(leftNums[i], rightNums[i])
	}
	fmt.Println("Part 1:", sum)
	rightMap := make(map[int]int)
	for _, num := range rightNums {
		rightMap[num] += 1
	}
	similiarity := 0
	for _, num := range leftNums {
		similiarity += rightMap[num] * num
	}
	fmt.Println("Part 2:", similiarity)
}
